ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageService.java

```diff
@@ -606,45 +605,52 @@ public void onCompletion(final BlobInfo routerResult, Exception routerException)
-                    Map<String, String> userMetadata = RestUtils.buildUserMetadata(routerResult.getUserMetadata());
-                    if (userMetadata == null) {
-                      restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, "application/octet-stream");
-                      restResponseChannel
-                          .setHeader(RestUtils.Headers.CONTENT_LENGTH, routerResult.getUserMetadata().length);
-                      response = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(routerResult.getUserMetadata()));
-                    } else {
-                      setUserMetadataHeaders(userMetadata, restResponseChannel);
-                      restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, 0);
-                      response = new ByteBufferReadableStreamChannel(AmbryBlobStorageService.EMPTY_BUFFER);
+          securityService
+              .processResponse(restRequest, restResponseChannel, routerResult.getBlobInfo(), new Callback<Void>() {
+                @Override
+                public void onCompletion(Void securityResult, Exception securityException) {
+                  securityCallbackTracker.markOperationEnd();
+                  ReadableStreamChannel response = null;
+                  boolean blobNotModified = restResponseChannel.getStatus() == ResponseStatus.NotModified;
+                  try {
+                    if (securityException == null) {
+                      if (subResource != null) {
+                        BlobInfo blobInfo = routerResult.getBlobInfo();
+                        Map<String, String> userMetadata = RestUtils.buildUserMetadata(blobInfo.getUserMetadata());
+                        if (userMetadata == null) {
+                          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_TYPE, "application/octet-stream");
+                          restResponseChannel
+                              .setHeader(RestUtils.Headers.CONTENT_LENGTH, blobInfo.getUserMetadata().length);
+                          response = new ByteBufferReadableStreamChannel(ByteBuffer.wrap(blobInfo.getUserMetadata()));
+                        } else {
+                          setUserMetadataHeaders(userMetadata, restResponseChannel);
+                          restResponseChannel.setHeader(RestUtils.Headers.CONTENT_LENGTH, 0);
+                          response = new ByteBufferReadableStreamChannel(AmbryBlobStorageService.EMPTY_BUFFER);
+                        }
+                      } else if (!blobNotModified) {
+                        response = routerResult.getBlobDataChannel();
+                      }
+                    }
+                  } catch (Exception e) {
+                    frontendMetrics.getSecurityResponseCallbackProcessingError.inc();
+                    securityException = e;
+                  } finally {
+                    if (response != null || securityException != null || blobNotModified) {
```

Comment:
i don't think you need this `if`. What case does it not cover?

Also note that the `submitResponse()` *has* to happen here. The old code had a check because it could potentially do another router call and so was not ready to submit the response. Not so in this case.